### ISSUE TITLE
SvelteKitAuth + refresh token rotation not working as expected

### ISSUE THREAD

**Issue by nicklasbondesson** [nextauthjs/next-auth#6447]
### Environment

```
System:
    OS: macOS 13.1
    CPU: (10) arm64 Apple M1 Pro
    Memory: 121.16 MB / 32.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 18.9.0 - ~/.nvm/versions/node/v18.9.0/bin/node
    Yarn: 1.22.19 - /opt/homebrew/bin/yarn
    npm: 8.19.1 - ~/.nvm/versions/node/v18.9.0/bin/npm
  Browsers:
    Chrome: 109.0.5414.87
    Safari: 16.2
```

### Reproduction URL

https://github.com/nicklasbondesson/sveltekitauth-token-rotation-demo

### Describe the issue

I'm trying to implement refresh token rotation (JWT) based on the guide over here https://authjs.dev/guides/basics/refresh-token-rotation using the [recently released SvelteKitAuth](https://vercel.com/blog/announcing-sveltekit-auth).

Once the access token has expired and is refreshed it seems that it is not remembered anymore, as shown below.

Here's the [entry point](https://github.com/nicklasbondesson/sveltekitauth-token-rotation-demo/blob/master/src/hooks.server.ts#L6) to the demo code where it can be reproduced.

![sveltekitauth-token-rotation-demo](https://user-images.githubusercontent.com/826299/213709641-3403c46d-fe94-43f1-8a50-fc7df84d0e12.gif)

### How to reproduce

Follow the README in here https://github.com/nicklasbondesson/sveltekitauth-token-rotation-demo

### Expected behavior

I think the expected behaviour would be that the refreshed token(s) would be remembered, just as it was on the initial login. If not, I would very much welcome some clarity around the matter.

**Comment by kai-ahmadov** [comment]
I've tried to follow the refresh token guideline as well, and I couldn't get the refreshed access_token to be persisted, which basically means that you have to refresh on every single requets. As far as I understand from the docs, on the server-side, the session is not persisted according to [this](https://authjs.dev/guides/basics/callbacks#session-callback). On the client side, it's persisted in a session cookie, but only once , on your initial login. Since you're refreshing from `hooks.server.ts`, I think you'd need to persist the session in a db

**Comment by half2me** [comment]
I'm having the same issue. When you use the `getSession()` function, it doesn't persist any new jwt cookie that it receives.
My workaround is to not use the `getSession()` and instead just use something like this in my `layout.server.ts`
```typescript
export const load: LayoutServerLoad = async ({ fetch }) => {
	const r = await fetch('/auth/session');
	const res = await r.json();
	console.log(res);
	return {
		session: res
	};
};
```

This actually persists the new jwt, so you don't keep refreshing the token on every request.
Also there is the case that your token can expire between page loads, and you might be relying on an access token to make an api call in the browser, but that token expires before you make a new call to your server to get a new jwt with the refreshed token.

So I like to also save the expiration time of the token, and schedule a function to run to invalidate it forcing a call to the server to refresh:
```typescript
setTimeout(() => {
  // invalidate session, force fetch for new token
  invalidate('/auth/session');
}, expiration);

**Comment by bfbay315** [comment]
@half2me Just curious where you added the timer to refresh the token?

**Comment by half2me** [comment]
> @half2me Just curious where you added the timer to refresh the token?

Just in my layout's onLoad()

**Comment by rbozan** [comment]
> I'm having the same issue. When you use the `getSession()` function, it doesn't persist any new jwt cookie that it receives. My workaround is to not use the `getSession()` and instead just use something like this in my `layout.server.ts`
> 
> ```ts
> export const load: LayoutServerLoad = async ({ fetch }) => {
> 	const r = await fetch('/auth/session');
> 	const res = await r.json();
> 	console.log(res);
> 	return {
> 		session: res
> 	};
> };
> ```
> 
> This actually persists the new jwt, so you don't keep refreshing the token on every request. Also there is the case that your token can expire between page loads, and you might be relying on an access token to make an api call in the browser, but that token expires before you make a new call to your server to get a new jwt with the refreshed token.
> 
> So I like to also save the expiration time of the token, and schedule a function to run to invalidate it forcing a call to the server to refresh:
> 
> ```ts
> setTimeout(() => {
>   // invalidate session, force fetch for new token
>   invalidate('/auth/session');
> }, expiration);
> ```

Sorry but does this work? It seems like a dirty hack

**Comment by half2me** [comment]
@rbozan works fine for me 🤷 . If there is any better way to do it, I'm all ears.

**Comment by bfbay315** [comment]
I'm doing something similar.  I'm not happy with it, but it's working.


In +layout.server.ts

```
export const load: LayoutServerLoad = async (event) => {
    return {
        session: await event.locals.getSession(),
    };
};

```
In layout.svelte

```
$: $page.data.session, refreshSession();

function refreshSession() {
	if (browser) {
		tokenExpirationMilliseconds = $page.data.session?.expires_in * 1000;
		if ($page.data.session?.error === "RefreshAccessTokenError") {
			logout()
		}else{
			if(!timer) {
				timer = setTimeout(() => {
					invalidateAll()
					timer = null;
				}, tokenExpirationMilliseconds)
			}
		}
	}
}
```

**Comment by half2me** [comment]
@bfbay315 yeah, in my case I only invalidate the auth endpoint instead of everything, but otherwise same idea. It works 🤷

**Comment by aakash14goplani** [comment]
I think this was recently addressed. You must be on or above v0.10.0 of `@auth/sveltekit` + you have to change `event.locals.getSession()` to `event.locals.auth()`. This new method updates cookies which helps resolving this issue. Unfortunately documentation is yet to be updated!

**Comment by rbozan** [comment]
> I think this was recently addressed - [refer](https://github.com/nextauthjs/next-auth/pull/9694). You must be on or above v0.10.0 of `@auth/sveltekit` + you have to change `event.locals.getSession()` to `event.locals.auth()`. This new method updates cookies which helps resolving this issue. Unfortunately documentation is yet to be updated!

Thanks for the tip. Just a note if someone else will try this, you also need the following for SveltekitAuth, otherwise the build will fail:

```
  basePath: "/api/auth",
```

**Comment by aakash14goplani** [comment]
@rbozan - where should we add this line?

**Comment by kinglouie** [comment]
currently testing out authjs and the new `event.locals.auth()` function successfully updates the session jwt token when reloading the page. I am current facing the issue that I don't know what would be the best way to update the client side session for token rotation purposes?
The `callbacks.session` callback from `SvelteKitAuth` only runs serverside, it would nice if there was a authjs function to achieve something similar clientside.

**Comment by aakash14goplani** [comment]
Finally, I found solution for this issue: https://blog.aakashgoplani.in/how-to-implement-refresh-token-rotation-in-sveltekitauth
